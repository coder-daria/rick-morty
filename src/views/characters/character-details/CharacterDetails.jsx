import React, { useMemo } from 'react';
import { instanceOf, shape, string } from 'prop-types';
import { Col, Divider } from 'antd';
import { v4 as uuid } from 'uuid';
import { map } from 'ramda';

import CharacterModel from '../model/CharacterModel';
import { StyledRow, CharacterImg } from './CharacterDetails.styles';

const NO_DATA_PLACEHOLDER = '-';

function CharacterDetails({ character }) {
  const characterDetails = useMemo(() => CharacterModel(character), [
    character,
  ]);

  return (
    <div>
      {map(item => {
        const isNameKey = item === 'name';
        const isImageKey = item === 'image';
        const value = characterDetails[item];

        return (
          <StyledRow key={uuid()}>
            {isNameKey && <span>{value || NO_DATA_PLACEHOLDER}</span>}
            {isImageKey && (
              <>
                <CharacterImg fallback={item} alt={item} src={value} />
                <Divider />
              </>
            )}
            {!isNameKey && !isImageKey && (
              <>
                <Col span={12}>{item}</Col>
                <Col span={12}>{value || NO_DATA_PLACEHOLDER}</Col>
              </>
            )}
          </StyledRow>
        );
      }, Object.keys(characterDetails))}
    </div>
  );
}

CharacterDetails.propTypes = {
  character: shape({
    created: string,
    episode: instanceOf(Array),
    gender: string,
    image: string,
    location: instanceOf(Object),
    name: string,
    origin: instanceOf(Object),
    species: string,
    status: string,
    type: string,
  }).isRequired,
};

export default CharacterDetails;
