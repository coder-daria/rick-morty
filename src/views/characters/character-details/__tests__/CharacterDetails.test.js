import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';

import {
  CHARACTER_DETAILED_MOCKED_DATA,
  CHARACTER_NOT_FORMATED_MOCKED_DETAILS,
} from '../../__mocks__/character-mock-data';

import CharacterDetails from '../CharacterDetails';

const setup = () =>
  shallow(
    <CharacterDetails
      character={{
        ...CHARACTER_DETAILED_MOCKED_DATA,
        ...CHARACTER_NOT_FORMATED_MOCKED_DETAILS,
      }}
    />,
  );

describe('<CharacterDetails />', () => {
  it('should match the snapshot', () => {
    const wrapper = setup();

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
