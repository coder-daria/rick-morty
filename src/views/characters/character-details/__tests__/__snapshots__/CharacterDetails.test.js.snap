// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterDetails /> should match the snapshot 1`] = `
<div>
  <Styled(Row)
    key="name"
  >
    <span>
      Rick Sanchez
    </span>
  </Styled(Row)>
  <Styled(Row)
    key="image"
  >
    <styled.img
      alt="image"
      src=""
    />
    <Divider />
  </Styled(Row)>
  <Styled(Row)
    key="created"
  >
    <Col
      span={12}
    >
      created
    </Col>
    <Col
      span={12}
    >
      11/25/2020, 4:59:08 PM
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="episode"
  >
    <Col
      span={12}
    >
      episode
    </Col>
    <Col
      span={12}
    >
      -
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="gender"
  >
    <Col
      span={12}
    >
      gender
    </Col>
    <Col
      span={12}
    >
      Male
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="location"
  >
    <Col
      span={12}
    >
      location
    </Col>
    <Col
      span={12}
    >
      Earth
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="origin"
  >
    <Col
      span={12}
    >
      origin
    </Col>
    <Col
      span={12}
    >
      Earth
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="species"
  >
    <Col
      span={12}
    >
      species
    </Col>
    <Col
      span={12}
    >
      Alien
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="status"
  >
    <Col
      span={12}
    >
      status
    </Col>
    <Col
      span={12}
    >
      Alive
    </Col>
  </Styled(Row)>
  <Styled(Row)
    key="type"
  >
    <Col
      span={12}
    >
      type
    </Col>
    <Col
      span={12}
    >
      -
    </Col>
  </Styled(Row)>
</div>
`;
